src/codegen/codegen_compatibility_visitor.cpp

```diff
@@ -0,0 +1,142 @@
+#include "visitors/lookup_visitor.hpp"
+
+namespace nmodl {
+namespace codegen {
+
+using visitor::AstLookupVisitor;
+
+/**
+ * \details Checks all the ast::AstNodeTypes that are not compatible with NMODL code
+ * generation and prints related messages. If there is some kind of incompatibility
+ * stop NMODL code generation.
+ */
+bool CodegenCompatibilityVisitor::find_incompatible_ast_nodes(Ast* node) {
+    std::vector<AstNodeType> incompatible_ast_types = {AstNodeType::SOLVE_BLOCK,
+                                                       AstNodeType::TERMINAL_BLOCK,
+                                                       AstNodeType::PARTIAL_BLOCK,
+                                                       AstNodeType::MATCH_BLOCK,
+                                                       AstNodeType::BA_BLOCK,
+                                                       AstNodeType::CONSTANT_BLOCK,
+                                                       AstNodeType::CONSTRUCTOR_BLOCK,
+                                                       AstNodeType::DESTRUCTOR_BLOCK,
+                                                       AstNodeType::DISCRETE_BLOCK,
+                                                       AstNodeType::FUNCTION_TABLE_BLOCK,
+                                                       AstNodeType::INDEPENDENT_BLOCK,
+                                                       AstNodeType::GLOBAL_VAR,
+                                                       AstNodeType::POINTER_VAR,
+                                                       AstNodeType::BBCORE_POINTER_VAR};
+    incompatible_ast_nodes = AstLookupVisitor().lookup(node, incompatible_ast_types);
+
+    std::stringstream ss;
+    for (auto it: incompatible_ast_nodes) {
+        if (it->is_solve_block()) {
+            auto solve_block = dynamic_cast<ast::SolveBlock*>(it.get());
+            auto method = solve_block->get_method();
+            if (method != nullptr && method->get_node_name() != "cnexp" &&
+                method->get_node_name() != "euler" && method->get_node_name() != "derivimplicit" &&
+                method->get_node_name() != "sparse") {
+                ss << "\"" << method->get_node_name() << "\" solving method used at [";
+                ss << method->get_token()->position() << "] not supported. ";
+                ss << "Supported methods are cnexp, euler, derivimplicit and sparse\n";
+            }
+        } else if (it->is_discrete_block() || it->is_partial_block()) {
+            std::string node_type_name = it->get_node_type_name();
+            // remove "Block" substring
+            node_type_name.erase(node_type_name.end() - 5, node_type_name.end());
+            // transform name of node type to capital letters, as it is in mod files
+            std::transform(node_type_name.begin(),
+                           node_type_name.end(),
+                           node_type_name.begin(),
+                           ::toupper);
+
```

Comment:
Name of the node not "necessarily" same as AST node name. Hence it would be better to have an extra method in the ast node like `get_nmodl_name` to return corresponding NMODL name. This method will be generated by jinja template if there is `nmodl` property exist in nmodl.yaml. 

@ohm314  can help if needed.

Reply:
I implemented this function using the `ast.hpp` template but I was wondering if all the nodes should have it, as there as some types that don't have `nmodl_name` at all or they have `None` as `nmodl_name`. I think that only the ones that have a useful `nmodl_name` should have them but I don't know if all the nodes should have this method for the sake of uniformity.

Reply:
I don't think they all need it. If it doesn't make sense, it should maybe not be there at all instead of returning placeholder values.